File: workspace/workflows/src/collection/traits.rs
use crate::activity::{Activity, ActivityContext, Error};
use async_trait::async_trait;
use serde_json::{json, Value};
use std::collections::HashSet;
use bosca_client::client::{Client, WorkflowJob};
use bosca_client::client::add_activity::ActivityInput;

pub struct CollectionTraitsActivity {
    id: String,
}

impl Default for CollectionTraitsActivity {
    fn default() -> Self {
        Self::new()
    }
}

impl CollectionTraitsActivity {
    pub fn new() -> CollectionTraitsActivity {
        CollectionTraitsActivity {
            id: "collection.traits.process".to_string(),
        }
    }
}

#[async_trait]
impl Activity for CollectionTraitsActivity {
    fn id(&self) -> &String {
        &self.id
    }

    fn create_activity_input(&self) -> ActivityInput {
        ActivityInput {
            id: self.id.to_owned(),
            name: "Process Traits".to_string(),
            description: "Process a Collection's Traits".to_string(),
            child_workflow_id: None,
            configuration: Value::Null,
            inputs: vec![],
            outputs: vec![],
        }
    }

    async fn execute(&self, client: &Client, _: &mut ActivityContext, job: &WorkflowJob) -> Result<(), Error> {
        if job.collection.is_none() {
            return Err(Error::new("missing collection".to_string()));
        }
        let collection = job.collection.as_ref().unwrap();
        let mut executed: HashSet<String>;
        if job.context.is_null() {
            executed = HashSet::new();
        } else {
            let ctx = job.context.as_object().unwrap();
            let value = ctx.get("executed").unwrap_or(&Value::Null);
            if value.is_null() {
                executed = HashSet::new();
            } else {
                executed = value
                    .as_array()
                    .unwrap()
                    .iter()
                    .map(|v| v.as_str().unwrap().to_string())
                    .collect();
            }
        }
        for trait_id in &collection.trait_ids {
            if executed.contains(trait_id) {
                continue;
            }
            let trait_ = client.get_trait(trait_id).await?;
            if trait_.is_none() {
                continue;
            }
            client
                .enqueue_child_workflows(&job.id.id, job.id.index, &job.id.queue, trait_.unwrap().workflow_ids)
                .await?;
            executed.insert(trait_id.clone());
            let updated_context = json!({"executed": executed.clone()});
            client
                .set_job_context(&job.id.id, job.id.index, &job.id.queue, &updated_context)
                .await?;
        }
        Ok(())
    }
}

File: workspace/runners/kotlin/activities/src/main/kotlin/io/bosca/workflow/collection/Traits.kt
package io.bosca.workflow.collection

import io.bosca.api.Client
import io.bosca.graphql.fragment.WorkflowJob
import io.bosca.graphql.type.ActivityInput
import io.bosca.workflow.Activity
import io.bosca.workflow.ActivityContext

class Traits(client: Client) : Activity(client) {

    override val id: String = ID

    override suspend fun toActivityDefinition(): ActivityInput {
        return ActivityInput(
            id = id,
            name = "Process Collection Traits",
            description = "",
            inputs = emptyList(),
            outputs = emptyList(),
        )
    }

    override suspend fun execute(context: ActivityContext, job: WorkflowJob) {
        val metadata = job.collection?.collection ?: error("missing collection")
        val executed = mutableSetOf<String>()
        if (job.context != null) {
            @Suppress("UNCHECKED_CAST")
            executed.addAll(((job.context as Map<*, *>)["executed"] as List<String>))
        }
        for (traitId in metadata.traitIds) {
            if (executed.contains(traitId)) continue
            val trait = client.workflows.getTrait(traitId) ?: continue
            client.workflows.enqueueChildWorkflows(
                trait.workflowIds,
                job.id
            )
            executed.add(traitId)
            client.workflows.setWorkflowJobContext(job.id, mapOf("executed" to executed.toList()))
        }
    }

    companion object {
        const val ID = "collection.traits.process"
    }
}